Add color and z-index to annotations, support multiclass output

jonasteuwen opened the pull request:
- Colors can now be added to annotations
- Z-indices can now be added to annotations to allow their sorting
- Sorting is now verified to work with ASAP, GeoJSON and Darwin V7
- Fixed a bug in polygons with multiple holes (see #224)
- Small changes in QuPath required to change the GeoJSON reading.
- `convert_annotations` now supports multiclass output
- A new annotation type: `AnnotationType.TAG`
- Ability to read Darwin V7 tags
- A bugfix for Darwin bounding boxes
- Automatic casting to AnnotationType.BOX for rectangular polygons
- Removal of function `rescale_geometry`

A todo for the future: Compress the GeoJSON output by using MultiPolygon

Fixes #188

moerlemans commented on tests/test_annotations.py:
The test for the holes from the other PR has disappeared. Shouldn't it be added here?

jonasteuwen replied:
There was a test already that checked the mask, I’ve added an extra hole

BPdeRooij commented on dlup/data/transforms.py:
```diff
@@ -79,12 +80,27 @@ def convert_annotations(
+                if curr_annotation.z_index is None and curr_annotation.label != roi_name:
+                    raise ValueError(f"Label {curr_annotation.label} has no z_index and no index_map is provided.")
+                assert curr_annotation.z_index
+                index_value = curr_annotation.z_index + 1
```
Is it appropriate to use z-index as an index value for the mask? Also, if a user has specified a specific z-index, should we add one to it? This is a DLUP-specific implementation, where normally index zero is reserved for "background". It would make more sense to use the z-index as is, if we use it at all.

jonasteuwen replied:
No, I've removed it.

BPdeRooij reviewed (commented):
In addition to the changes described above, this PR also implements:
- A new annotation type: `AnnotationType.TAG`
- Ability to read Darwin V7 tags
- A bugfix for Darwin bounding boxes
- Automatic casting to AnnotationType.BOX for rectangular polygons
- Removal of function `rescale_geometry`

Overall the changes are a good addition and the simplification of the code for annotations is welcome. The color and z-index properties are useful properties to add to annotations.

I have the following questions and points for discussion:
- Is the new annotation type `AnnotationType.TAG` supposed to replace or supplement the existing `labels` for DLUP datasets? Tags are currently stored in the `WsiAnnotations`, but not accessed by the dataset object. Labels are key-value pairs, but tags only support a single value (which can be a problem for Darwin tags). An annotations type `AnnotationType.LABEL` with a key-value pair would combine these two. If only a single value is given, a True value can be attributed as pair to the key.
- Darwin tags can have sub-tags, which are stored as attributes. These attributes can contain one or more strings. The attributes are not taken into consideration now. An example of this can be found below: 
```
    {
      "id": "XXX",
      "name": "diagnosable slide",
      "properties": [],
      "slot_names": [
        "0"
      ],
      "tag": {}
    },
    {
      "attributes": [
        "optimal staining"
      ],
      "id": "XXX",
      "name": "staining quality",
      "properties": [],
      "slot_names": [
        "0"
      ],
      "tag": {}
    }
```

- (See comment in code as well) Are z-indices an appropriate substitution for an index map? Z-indices serve as an ordering tools and while it can be correct to use as an index value in some cases, there are also cases where it will lead to errors. 
- (See comment in code as well) The automatic casting of rectangular polygons to `AnnotationType.BOX` can lead to unintended behavior. Would it be a solution to add an additional property to explicitly cast to `AnnotationType.BOX`
- (See comment in code as well) What is the reason behind removing the `rescale_geometry` function and scaling factor from GeoJSON annotations.
- (See comment in code as well) `FeatureCollection` object cannot have a "properties" member according to GeoJSON documentation. 
- (See comment in code as well) Both "color" and "z-index" can be added to `Feature` "properties" directly. Is there a specific reason for adding it to "classification"? (Perhaps to be in agreement with the QuPath format?)

jonasteuwen commented on dlup/annotations.py:
```diff
@@ -813,41 +949,13 @@ def as_geojson(self) -> GeoJsonDict:
-                else:
-                    annotation = shapely.geometry.MultiPoint(annotation_list)
-                json_dict = _geometry_to_geojson(annotation, label=label)
+        data: GeoJsonDict = {"type": "FeatureCollection", "properties": None, "features": [], "id": None}
```
Thanks, that's a good addition. It's something which would be dlup-only so we need to handle this carefully and try to avoid too much custom properties. I would propose to have a dlup-xml format which can support this better.

jonasteuwen commented on setup.py:
```diff
@@ -17,7 +17,7 @@
 
 install_requires = [
-    "numpy>=1.26.0",
+    "numpy==1.26.4",
```
@BPdeRooij Yes, the newer versions of numpy give issues because they have deprecated some things, in dlup v1.0 we can see how to do that

BPdeRooij replied:
This is a good issue to describe and keep track. Perhaps, we can already give a range to the numpy version according to the third example here: https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/

jonasteuwen commented on dlup/data/transforms.py:
```diff
@@ -79,12 +80,27 @@ def convert_annotations(
+            and curr_annotation.annotation_type == AnnotationType.BOX
+        ):
             min_x, min_y, max_x, max_y = curr_annotation.bounds
             boxes[curr_annotation.label].append(((int(min_x), int(min_y)), (int(max_x - min_x), int(max_y - min_y))))
```
Boxes are removed because intersections of boxes with boxes are polygons. If we want that behavior we need a transform to do this explicitly

BPdeRooij replied:
Interesting point. This would be the case one of the boxes would be rotated with respect to the other box.

jonasteuwen commented on dlup/annotations.py:
```diff
@@ -156,118 +327,158 @@ def __str__(self) -> str:
-        ]
-
+        _coordinates = coordinates["coordinates"]
+        annotation_type = AnnotationType.BOX if _is_rectangle(_coordinates) else AnnotationType.POLYGON
```
This is how QuPath parses it, and it's also needed to be able to go from V7 Darwin <-> GeoJSON. It would mean we would break that behavior. I would be in favor of breaking it here and implement our own annotation format that can support this.

It doesn't lead to undefined behavior. With the new changes in `convert_annotations` it still behaves well as the BOX is also a `dlup.annotations.Polygon`. It might be good to have it as a box everywhere, it could be useful for some checks (but which?).

What do you think?

BPdeRooij replied:
This is no longer an issue with the new changes in `convert_annotations`. If boxes are removed from the `convert_annotations`, boxes will be correctly incorporated in the mask. 

I do think we should be thoughtful with how we handle boxes. One use case would be object detection with (non-rotated) bounding boxes as targets. A user can define a custom `convert_annotations` function, but it might be useful to already provide boxes in the standard `"annotation_data"` dictionary. What do you think would be a good solution for this use case? Perhaps we can introduce a more restrictive annotation type `AnnotationType.BOUNDINGBOX`, which only allows non-rotated boxes.

jonasteuwen replied:
Once we have data with clear boxes, I think we can do something with this to see how to handle those. I would for instance think about an object BoundingBox that returns a tuple[int, int, int, int] rather than directly having these as Polygons.

I just don't have annotations like that yet. Since things we consider are usually rotationally invariant we often use points

jonasteuwen commented on dlup/annotations.py:
```diff
@@ -288,151 +501,81 @@ def _geometry_to_geojson(geometry: Polygon | Point, label: str) -> dict[str, Any
-    def label(self) -> str:
-        """The label name for this annotation."""
-        return self._label
+    if color is not None:
```
It's how QuPath parses it unfortunately. How do you propose how we do this?

BPdeRooij replied:
In that case it would be good to follow the convention of QuPath to keep it compatible. It is still correct according to the GeoJSON documentation.

jonasteuwen commented on dlup/annotations.py:
```diff
@@ -156,118 +327,158 @@ def __str__(self) -> str:
         return f"{self.annotation_class}, {self.wkt}"
 
 
-def rescale_geometry(geometry: Union[Point, Polygon], scaling: float | None = None) -> Union[Point, Polygon]:
-    if scaling is None:
-        return geometry
-    if scaling == 1.0:
-        return geometry
-
-    scaled_geometry = shapely.affinity.scale(geometry, scaling, scaling)
-    if isinstance(geometry, Polygon):
-        return Polygon(scaled_geometry, a_cls=geometry.annotation_class)
-
-    if isinstance(geometry, Point):
-        return Point(scaled_geometry, a_cls=geometry.annotation_class)
-
-    raise AnnotationError(f"geometry type {type(geometry)} is not a valid dlup type.")
-
-
```
It can be useful, but I do not see a proper use case for this. I think you should always save the coordinates in level 0. The reason why it is there for ASAP is because they sometimes give them together with images resampled to 0.5um/pixel. If you want to read them in the native resolution you would need a scaling factor. 

It also behaved a bit erratic

BPdeRooij replied:
I was thinking of the use case when working with down-scaled data or feature vectors generated by models. For example, in [ahcore's _ValidationDataset](https://github.com/NKI-AI/ahcore/blob/a56f12326aec44b4c15bb1412813d6e89b310bfd/ahcore/utils/callbacks.py#L30) or with the ahcore's ZarrReader backend, the annotations are not always in level 0. Rescaling the geometry of `WsiAnnotations` might be necessary in these cases.

jonasteuwen replied:
Yes, but those are masks, so that’s not a problem there. If we save, we can always rescale when we save

jonasteuwen commented:
Thank you @BPdeRooij. I think I've addressed most of your questions in replies to your specific questions. Let me know if these are sufficient

- Regarding not having 1-1 support of GeoJSON, thently such that QuPath can read it. I've added this to the docstring now. Perhaps in a future version, we can add `strict` as a parameter to adhere strictly to the GeoJSON standard export or  use`qupath_compatible`. The current implementation also has tests associated with so we can check if the reading with another reader (specifically Darwin to GeoJSON) is still working. I think for the purpose of the tests this is required.  I think that in a future version we should have our own serialization format, or look at the cvat one or the darwin one for inspiration
- Regarding the TAG: consider the current version as beta. Indeed, hierarchical tags seem to be possible, I'd like to implement those properly in 1.0. Currently you can read a Darwin V7 file without crashing and the Polygon and Points are processed correctly.
- A tag is an image-level label. I think a LABEL is a specific value associated to a region (such as "has tissue"). I have yet to come across a good use case where we use this, so it is not really clear to me yet how to do that
- numpy is a difficult story as newer versions seem to break older version, this needs to be investigated in more detail. I also don't want to require numpy version too new as other dependencies break.

I feel we can have these comments as specific issues that would be relevant for a dlup v1.0. Let me know what you think

BPdeRooij reviewed (dismissed):
Thank you for your changes and responses to the questions. The ability of creating multiclass output is a nice addition as well. We should keep track of the aforementioned points relating to numpy versioning, (bounding) boxes and new annotation type.

The changes and additional features are motivated sufficiently and this PR is good to merge.
